### PR TITLE
Allow logger to accept a device

### PR DESCRIPTION
**Pull request by davydovanton**
This is first step for all logger issues like https://github.com/hanami/hanami/issues/481.

/cc @jodosha

### CODE REVIEW

**Comment by davydovanton** [comment]
@hanami/core-team could you check this PR? I can't complete all logger issues without this :disappointed:

**Comment by jodosha** [comment]
@davydovanton Hey, thanks for this PR. :smile: 

There are a couple of improvements to the tests and a more use cases to cover.

Tests:

  * The log file shouldn't be added to the source code.
  * We should use a `tmp/` directory.
  * Ensure that the file is removed in a `before` block
  * When testing an already existing file we can use `FileUtils.touch` to create it under `tmp/`
  * We should assert that the UNIX permissions are the same that `Logger` (from stdlib) uses.

Features:

  * Accept a relative path
  * Accept `Pathname`
  * Accept `IO`/`StringIO`
  * Make `#close` safe when the internal device is `STDOUT` (see below).

---

### STDOUT close

```ruby
require 'logger'

logger = Logger.new(STDOUT)
logger.close

puts "hi"

__END__

➜ ruby logging.rb
logging.rb:6:in `write': closed stream (IOError)
	from logging.rb:6:in `puts'
	from logging.rb:6:in `puts'
	from logging.rb:6:in `<main>'
```

**Comment by davydovanton** [comment]
@jodosha thanks for review. I have some questions:
1. What do you expect with `Pathname`?
2. I check `IO` in `File.new` test. Do you want to check pure `IO` class too?

Thanks!

**Comment by jodosha** [comment]
@davydovanton Thanks for the progress :+1: 

> 1. What do you expect with Pathname?

Same values that we expect from a path represented as a string: relative and absolute path.

> 2. I check IO in File.new test. Do you want to check pure IO class too?

Yes please because `File` is backed by filesystem and it was mandatory to test. While `IO` or `StringIO` are really nice to have because useful for testing purposes or for extensibility with other third party gems.

**Comment by davydovanton** [comment]
@jodosha done. I added tests for `IO` and `Pathname` class too.

File: lib/hanami/logger.rb [hunk 1]
```diff
@@ -122,6 +133,13 @@ def application_name
       @application_name || _application_name_from_namespace || _default_application_name
     end
 
+    # Close the logging device if this device isn't an STDOUT
+    #
+    # @since x.x.x
+    def close
+      super if @log_device != STDOUT
```
Comment (jodosha) [inline lib/hanami/logger.rb]: @davydovanton We should also guard against `$stdout`.

```ruby
super unless [STDOUT, $stdout].include?(@log_device)
```

Comment (davydovanton) [inline lib/hanami/logger.rb]: yep, you're right. Thanks for catch! :+1:

**Comment by davydovanton** [comment]
@jodosha done :+1:

**Comment by jodosha** [comment]
@davydovanton Thank you. Merged :+1:

**Comment by davydovanton** [comment]
@jodosha thanks for merge and review :+1:

**Pull request merged**